src/polysolve/nonlinear/Solver.cpp

```diff
@@ -0,0 +1,388 @@
+        solver_info["line_search"] = line_search_name;
+    }
+
+    void Solver::minimize(Problem &objFunc, TVector &x)
+    {
+        constexpr double NaN = std::numeric_limits<double>::quiet_NaN();
+
+        // ---------------------------
+        // Initialize the minimization
+        // ---------------------------
+
+        reset(x.size()); // place for children to initialize their fields
+
+        TVector grad = TVector::Zero(x.rows());
+        TVector delta_x = TVector::Zero(x.rows());
+
+        // double factor = 1e-5;
+
+        // Set these to nan to indicate they have not been computed yet
+        double old_energy = NaN;
+
+        {
+            POLYSOLVE_SCOPED_TIMER("constraint set update", constraint_set_update_time);
+            objFunc.solution_changed(x);
+        }
+
+        {
+            POLYSOLVE_SCOPED_TIMER("compute gradient", grad_time);
+            objFunc.gradient(x, grad);
+        }
+        double first_grad_norm = compute_grad_norm(x, grad);
+        if (std::isnan(first_grad_norm))
+        {
+            this->m_status = cppoptlib::Status::UserDefined;
+            m_error_code = ErrorCode::NAN_ENCOUNTERED;
+            log_and_throw_error(m_logger, "[{}] Initial gradient is nan; stopping", name());
+            return;
+        }
+        this->m_current.xDelta = NaN; // we don't know the initial step size
+        this->m_current.fDelta = old_energy;
+        this->m_current.gradNorm = first_grad_norm / (normalize_gradient ? first_grad_norm : 1);
+
+        const auto current_g_norm = this->m_stop.gradNorm;
+        this->m_stop.gradNorm = first_grad_norm_tol;
+        this->m_status = checkConvergence(this->m_stop, this->m_current);
+        if (this->m_status != cppoptlib::Status::Continue)
+        {
+            POLYSOLVE_SCOPED_TIMER("compute objective function", obj_fun_time);
+            this->m_current.fDelta = objFunc.value(x);
+            m_logger.info(
+                "[{}] Not even starting, {} (f={:g} ‖∇f‖={:g} g={:g} tol={:g})",
+                name(), this->m_status, this->m_current.fDelta, first_grad_norm, this->m_current.gradNorm, this->m_stop.gradNorm);
+            update_solver_info(this->m_current.fDelta);
+            return;
+        }
+        this->m_stop.gradNorm = current_g_norm;
+
+        // TODO
+        // utils::Timer timer("non-linear solver", this->total_time);
+        // timer.start();
+
+        if (m_line_search)
+            m_line_search->use_grad_norm_tol = use_grad_norm_tol;
+
+        objFunc.save_to_file(x);
+
+        m_logger.debug(
+            "Starting {} solve f₀={:g} ‖∇f₀‖={:g} "
+            "(stopping criteria: max_iters={:d} Δf={:g} ‖∇f‖={:g} ‖Δx‖={:g})",
+            name(), objFunc.value(x), this->m_current.gradNorm, this->m_stop.iterations,
+            this->m_stop.fDelta, this->m_stop.gradNorm, this->m_stop.xDelta);
+
+        update_solver_info(objFunc.value(x));
+
+        do
+        {
+            if (name() == "MMA")
```

Comment:
this is bad

Reply:
@Huangzizhou can this be refactored somehow? why is this necessary? same for the other one